### DISCUSSION TITLE
部署常见问题说明

### DISCUSSION THREAD

**Discussion by Jinnrry** [Jinnrry/PMail#170] [Q&A]
### 1、服务器收信发信失败，日志中出现timeout相关报错
此类错误请先检查自己服务器网络是否正常，使用telnet、nslookup之类的命令排除

### 2、服务器25端口被封
大部分VPS提供商（比如阿里云、腾讯云、AWS等）默认是封禁25端口的，需要提工单单独申请解封。https://github.com/Jinnrry/PMail/issues/160

### 3、25端口被封能否使用其他端口？比如465
不能！ 详见：https://github.com/Jinnrry/PMail/issues/130

### 4、服务器80 443端口有其他服务占用，能否使用？
可以，https://github.com/Jinnrry/PMail/discussions/59 可以参考其他用户写的博客

### 5、自己申请证书要匹配哪些域名？
根域名、pop.[域名]、smtp.[域名]

### 6、邮件发送不出去，报PTR错误
<details><pre><code>
邮件投递失败550 5.7.25 [你的ip] The IP address sending this message does not
5.7.25 have a PTR record setup, or the corresponding forward DNS entry does
5.7.25 not match the sending IP. As a policy, Gmail does not accept
5.7.25 messages from IPs with missing PTR records. For more information, go
5.7.25 to
5.7.25 https://support.google.com/a?p=sender-guidelines-ip
5.7.25 To learn more about Gmail requirements for bulk senders, visit
5.7.25 https://support.google.com/a?p=sender-guidelines. 
</code></pre></details>

[使用这个工具](https://powerdmarc.com/zh/spf-record-lookup/)检查域名DNS记录中的SPF设置，常见错误是服务器有ipv6地址，DNS记录里面只写了ipv4地址。通讯的使用又使用了ipv6地址。 #181

### 7、中文字体乱码
检查服务器是否安装中文字体、检查数据库字符集、检查数据表字符集、检查数据库连接参数字符集
https://github.com/Jinnrry/PMail/issues/186

### 8、客户端设置

<details><pre><code>

部分客户端自动设置不能正确设置POP和SMTP服务器连接信息（表现为无法收信、或者无法发信），因此需要手动设置，以IOS邮件APP为例，首先选择POP协议，然后依次填入邮箱地址，pop信息（主机名为pop服务域名，即 `pop.[你的域名]`），smtp信息(主机名为smtp服务域名，即`smtp.[你的域名]`)，用户名密码都不能为空

![img](https://raw.githubusercontent.com/Jinnrry/PMail/refs/heads/master/docs/settings.jpg)

</code></pre></details>

### 9、邮件进垃圾箱

https://github.com/Jinnrry/PMail/issues/226
使用类似 [spamhaus](https://check.spamhaus.org/) 的服务检查IP是否干净

### 10、出现很多发件人是自己的邮件

这些邮件是其他人伪造的，并不是PMail漏洞，也不属于bug #250 

### 11、不设置密码也能发件

https://github.com/Jinnrry/PMail/issues/255

实际上只是服务端当成收件处理了而已

### 12、有很多垃圾邮件怎么办？

<details><pre><code>

目前支持2种防垃圾方案
1、配置文件中  `spamFilterLevel`设置为1或2，这个配置表示垃圾邮件过滤级别，0不过滤、1 spf dkim 校验均失败时过滤，2 spf校验不通过时过滤，这个设置是基于SMTP协议进行过滤的，如果对方服务器完全符合spf和dkim校验，那么是过滤不了的

2、基于机器学习的过滤方案，https://github.com/Jinnrry/PMail/blob/master/README_CN.md

</code></pre></details>

**Comment by Jinnrry** [comment]
还有其他部署问题可以在这里提问

**Reply by suyizegg** [reply]
at /work/services/setup/db.go:64 at /work/services/setup/db.go:120 invalid character ':' after top-level value
老师，这个没找到在哪

**Reply by Jinnrry** [reply]
> at /work/services/setup/db.go:64 at /work/services/setup/db.go:120 invalid character ':' after top-level value 老师，这个没找到在哪

数据库连接信息不对吧

**Reply by hanzemo** [reply]
<img width="1919" height="933" alt="image" src="https://github.com/user-attachments/assets/99cbdabf-59f2-4a63-b1f2-a3be7a6bed11" />
一直卡在这样

**Reply by Jinnrry** [reply]
> <img alt="image" width="1919" height="933" src="https://private-user-images.githubusercontent.com/145012075/473766896-99cbdabf-59f2-4a63-b1f2-a3be7a6bed11.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQyNzg2NzEsIm5iZiI6MTc1NDI3ODM3MSwicGF0aCI6Ii8xNDUwMTIwNzUvNDczNzY2ODk2LTk5Y2JkYWJmLTU5ZjItNGE2My1iMWYyLWEzYmU3YTZiZWQxMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgwNFQwMzMyNTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01Mzk3ZWNiNzQ2NjJkYjg2YWY0NGM1NzE0ZDQ2YjIzYTc0YWMyMWEyODgxMzk0Y2QwNDY3MmQ3MTE5NTUwMDhjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.RU3pWAR-w2yq5tcuHVUQnDU50YSrP2bJNvCKALFWXr8"> 一直卡在这样

看下命令行输出，应该是证书签发等待时间较长。你如果使用DNS方式验证的话，DNS记录需要你自己配置，涉及DNS缓存，因此耗时会比较长

**Comment by henji521** [comment]
老师,问下能正常发送,但是不能收件是为什么呢?

**Reply by henji521** [reply]
25端口必须要打开码?

**Reply by Jinnrry** [reply]
25端口必须打开
1.检查端口是不是通的
2.检查dns记录配置对了没

**Reply by henji521** [reply]
嗯嗯对了老师,已经解决了

**Comment by henji521** [comment]
在问下老师 多域名如何部署呀 已经部署了添加的话

**Reply by Jinnrry** [reply]
目前还没考虑过这种需求。你可以把新的域名加到配置文件的`domains`项里面去，然后手动更新ssl证书

**Comment by Wu-HZ** [comment]
我有一个疑问，能收件，代表25端口被访问到了么

**Reply by Jinnrry** [reply]
是的，代表25端口的入站是通的

**Reply by Wu-HZ** [reply]
🤣看来阿里云是只开放了25的入站，我发件是投递失败的

**Reply by Jinnrry** [reply]
发件的报错是网络超时或者连接失败类的吗

**Reply by Wu-HZ** [reply]
> 发件的报错是网络超时或者连接失败类的吗

运行这个项目的时候我没找到日志出现的位置，后因为看到25端口被封用不了所以把这个项目从服务器上删除了，不能够再复现了抱歉。

**Reply by Jinnrry** [reply]
哦哦，没事。看其他人的回复，阿里云可以提工单找客服解封

**Reply by hetaoawa** [reply]
> 发件的报错是网络超时或者连接失败类的吗

我这边也有这种问题了（
收件正常，发件会报
`[error][2025-01-03 20:56:56][ac11sdgdf][/work/utils/send/send.go:213][0xc0002bcda0] 邮件投递失败dial tcp 157.255.4.243:25: i/o timeout
[error][2025-01-03 20:56:56][/work/hooks/framework/framework.go:123]params error json: cannot unmarshal object into Go struct field HookDTO.ErrMap of type error
[error][2025-01-03 20:56:56][/work/hooks/framework/framework.go:123]params error json: cannot unmarshal object into Go struct field HookDTO.ErrMap of type error
[error][2025-01-03 20:56:56][/work/hooks/framework/framework.go:123]params error json: cannot unmarshal object into Go struct field HookDTO.ErrMap of type error
`
感觉像是出站被封了，我交个工单试试

upd. 售后机器人说不支持解禁25端口，遂放弃（sad

**Comment by DavidYC1212** [comment]
请问 自动部署SSL两种验证方式都一直转圈圈是怎么回事 这里是日志：
2024/09/15 20:27:39 [INFO] acme: Registering account for i@xxxxx.com
[info][2024-09-15 20:27:39][/home/runner/work/PMail/PMail/server/services/setup/ssl/ssl.go:198]wait ssl
2024/09/15 20:27:39 [INFO] [pmail.xxxxx.com, smtp.xxxxx.com, pop.xxxxx.com] acme: Obtaining bundled SAN certificate
2024/09/15 20:27:41 [INFO] [pmail.xxxxx.com] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz-v3/403952487196
2024/09/15 20:27:41 [INFO] [pop.xxxxx.com] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz-v3/403952487206
2024/09/15 20:27:41 [INFO] [smtp.xxxxx.com] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz-v3/403952487216
2024/09/15 20:27:41 [INFO] [pmail.xxxxx.com] acme: Could not find solver for: tls-alpn-01
2024/09/15 20:27:41 [INFO] [pmail.xxxxx.com] acme: Could not find solver for: http-01
2024/09/15 20:27:41 [INFO] [pmail.xxxxx.com] acme: use dns-01 solver
2024/09/15 20:27:41 [INFO] [smtp.xxxxx.com] acme: Could not find solver for: tls-alpn-01
2024/09/15 20:27:41 [INFO] [smtp.xxxxx.com] acme: Could not find solver for: http-01
2024/09/15 20:27:41 [INFO] [smtp.xxxxx.com] acme: use dns-01 solver
2024/09/15 20:27:41 [INFO] [pop.xxxxx.com] acme: Could not find solver for: tls-alpn-01
2024/09/15 20:27:41 [INFO] [pop.xxxxx.com] acme: Could not find solver for: http-01
2024/09/15 20:27:41 [INFO] [pop.xxxxx.com] acme: use dns-01 solver
2024/09/15 20:27:41 [INFO] [pmail.xxxxx.com] acme: Preparing to solve DNS-01
[info][2024-09-15 20:27:42][/home/runner/work/PMail/PMail/server/services/setup/ssl/challenge.go:63]Presenting challenge Info : {FQDN:_acme-challenge.pmail.xxxxx.com. EffectiveFQDN:_acme-challenge.pmail.xxxxx.com. Value:UpAf5UJj4tKyMqy8dld5A91Cdst34C_QcNg76prq2hs}
[info][2024-09-15 20:27:42][/home/runner/work/PMail/PMail/server/services/setup/ssl/challenge.go:69]SSL Log:pmail.xxxxx.com yoIIpvJlrF5CBAQMSaL3Jf3rPTmq29p5po6byGDbY38 yoIIpvJlrF5CBAQMSaL3Jf3rPTmq29p5po6byGDbY38.g_PCMk_0bBYgB-BEbOFktn2yV21Qhc1eIsiLPBTNTpo
2024/09/15 20:27:42 [INFO] [smtp.xxxxx.com] acme: Preparing to solve DNS-01
[info][2024-09-15 20:27:43][/home/runner/work/PMail/PMail/server/services/setup/ssl/challenge.go:63]Presenting challenge Info : {FQDN:_acme-challenge.smtp.xxxxx.com. EffectiveFQDN:_acme-challenge.smtp.xxxxx.com. Value:dxH79Ilztj0IE2qEyZZYAEtyntS-LrmkjxBnDv16ilo}
[info][2024-09-15 20:27:43][/home/runner/work/PMail/PMail/server/services/setup/ssl/challenge.go:69]SSL Log:smtp.xxxxx.com eg7SlJAUxpVmKnXINhkbqg0Zjj_E3iq4WYDs1fhx_Po eg7SlJAUxpVmKnXINhkbqg0Zjj_E3iq4WYDs1fhx_Po.g_PCMk_0bBYgB-BEbOFktn2yV21Qhc1eIsiLPBTNTpo
2024/09/15 20:27:43 [INFO] [pop.xxxxx.com] acme: Preparing to solve DNS-01
[info][2024-09-15 20:27:43][/home/runner/work/PMail/PMail/server/services/setup/ssl/challenge.go:63]Presenting challenge Info : {FQDN:_acme-challenge.pop.xxxxx.com. EffectiveFQDN:_acme-challenge.pop.xxxxx.com. Value:jDP2qbVyZ9Mhj6Kc_wzxbewCgb3otX3Uhu1y7tLmeQg}
[info][2024-09-15 20:27:43][/home/runner/work/PMail/PMail/server/services/setup/ssl/challenge.go:69]SSL Log:pop.xxxxx.com Xxu3RcJTFCt5q8bmwtaa0gCsudC3qsXnWKlcH5W9XPw Xxu3RcJTFCt5q8bmwtaa0gCsudC3qsXnWKlcH5W9XPw.g_PCMk_0bBYgB-BEbOFktn2yV21Qhc1eIsiLPBTNTpo
2024/09/15 20:27:43 [INFO] [pmail.xxxxx.com] acme: Trying to solve DNS-01
2024/09/15 20:27:44 [INFO] [pmail.xxxxx.com] acme: Checking DNS record propagation. [nameservers=google-public-dns-a.google.com:53,google-public-dns-b.google.com:53]
2024/09/15 20:27:49 [INFO] Wait for propagation [timeout: 1h0m0s, interval: 5s]
2024/09/15 20:27:51 [INFO] [pmail.xxxxx.com] acme: Waiting for DNS record propagation.
2024/09/15 20:27:57 [INFO] [pmail.xxxxx.com] acme: Waiting for DNS record propagation.
2024/09/15 21:28:04 [INFO] retry due to: acme: error: 400 :: POST :: https://acme-v02.api.letsencrypt.org/acme/chall-v3/403952487196/hEsf2Q :: urn:ietf:params:acme:error:badNonce :: JWS has an invalid anti-replay nonce: "UAMvsxnwAKGLHVnuLLXIIcdwo71HhFmtHSdbi6aGTPYKeHh-4is"
2024/09/15 21:28:10 [INFO] [pmail.xxxxx.com] The server validated our request
2024/09/15 21:28:10 [INFO] [smtp.xxxxx.com] acme: Trying to solve DNS-01
2024/09/15 21:28:11 [INFO] [smtp.xxxxx.com] acme: Checking DNS record propagation. [nameservers=google-public-dns-a.google.com:53,google-public-dns-b.google.com:53]
2024/09/15 21:28:16 [INFO] Wait for propagation [timeout: 1h0m0s, interval: 5s]
2024/09/15 21:32:53 [INFO] acme: Trying to resolve account by key
[info][2024-09-15 21:32:53][/home/runner/work/PMail/PMail/server/services/setup/ssl/ssl.go:198]wait ssl
2024/09/15 21:32:53 [INFO] [pmail.xxxxx.com, smtp.xxxxx.com, pop.xxxxx.com] acme: Obtaining bundled SAN certificate
2024/09/15 21:32:55 [INFO] [pmail.xxxxx.com] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz-v3/403952487196
2024/09/15 21:32:55 [INFO] [pop.xxxxx.com] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz-v3/403952487206
2024/09/15 21:32:55 [INFO] [smtp.xxxxx.com] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz-v3/403952487216
2024/09/15 21:32:55 [INFO] [pmail.xxxxx.com] acme: authorization already valid; skipping challenge
2024/09/15 21:32:55 [INFO] [pop.xxxxx.com] acme: Could not find solver for: tls-alpn-01
2024/09/15 21:32:55 [INFO] [pop.xxxxx.com] acme: Could not find solver for: http-01
2024/09/15 21:32:55 [INFO] [pop.xxxxx.com] acme: use dns-01 solver
2024/09/15 21:32:55 [INFO] [smtp.xxxxx.com] acme: Could not find solver for: tls-alpn-01
2024/09/15 21:32:55 [INFO] [smtp.xxxxx.com] acme: Could not find solver for: http-01
2024/09/15 21:32:55 [INFO] [smtp.xxxxx.com] acme: use dns-01 solver
2024/09/15 21:32:55 [INFO] [pop.xxxxx.com] acme: Preparing to solve DNS-01
[info][2024-09-15 21:32:56][/home/runner/work/PMail/PMail/server/services/setup/ssl/challenge.go:63]Presenting challenge Info : {FQDN:_acme-challenge.pop.xxxxx.com. EffectiveFQDN:_acme-challenge.pop.xxxxx.com. Value:jDP2qbVyZ9Mhj6Kc_wzxbewCgb3otX3Uhu1y7tLmeQg}
[info][2024-09-15 21:32:56][/home/runner/work/PMail/PMail/server/services/setup/ssl/challenge.go:69]SSL Log:pop.xxxxx.com Xxu3RcJTFCt5q8bmwtaa0gCsudC3qsXnWKlcH5W9XPw Xxu3RcJTFCt5q8bmwtaa0gCsudC3qsXnWKlcH5W9XPw.g_PCMk_0bBYgB-BEbOFktn2yV21Qhc1eIsiLPBTNTpo
2024/09/15 21:32:56 [INFO] [smtp.xxxxx.com] acme: Preparing to solve DNS-01
[info][2024-09-15 21:32:56][/home/runner/work/PMail/PMail/server/services/setup/ssl/challenge.go:63]Presenting challenge Info : {FQDN:_acme-challenge.smtp.xxxxx.com. EffectiveFQDN:_acme-challenge.smtp.xxxxx.com. Value:dxH79Ilztj0IE2qEyZZYAEtyntS-LrmkjxBnDv16ilo}
[info][2024-09-15 21:32:56][/home/runner/work/PMail/PMail/server/services/setup/ssl/challenge.go:69]SSL Log:smtp.xxxxx.com eg7SlJAUxpVmKnXINhkbqg0Zjj_E3iq4WYDs1fhx_Po eg7SlJAUxpVmKnXINhkbqg0Zjj_E3iq4WYDs1fhx_Po.g_PCMk_0bBYgB-BEbOFktn2yV21Qhc1eIsiLPBTNTpo
2024/09/15 21:32:56 [INFO] [pop.xxxxx.com] acme: Trying to solve DNS-01
2024/09/15 21:32:56 [INFO] [pop.xxxxx.com] acme: Checking DNS record propagation. [nameservers=google-public-dns-a.google.com:53,google-public-dns-b.google.com:53]
2024/09/15 21:33:01 [INFO] Wait for propagation [timeout: 1h0m0s, interval: 5s]
这里是DNS：
![image](https://github.com/user-attachments/assets/d027dae6-a688-482e-9930-729314115c55)
此日志显示试了几次DNS，但我之前试另一种也是一直转圈
已阅读readme和issues但没有类似问题
使用腾讯云轻量服务器

**Reply by DavidYC1212** [reply]
等了两个小时证书发下来了……已经解决了

**Reply by Jinnrry** [reply]
DNS因为有缓存，需要等待较长时间是正常的

**Comment by askthesea** [comment]
遇到一个特别奇怪的问题，按照教程在debian上部署好之后，开始几分钟内，一切正常，但是很快就会发现部署的邮件网页（pmail.xxx.com）不可以访问，客户端邮件无法收取，此时再次运行部署教程最后一步的命令“docker run -p 25:25 -p 80:80 -p 443:443 -p 110:110 -p 465:465 -p 995:995 -v $(pwd)/config:/work/config ghcr.io/jinnrry/pmail:latest”，一切又恢复正常几分钟，然后再次陷入死循环，问了服务器官方，此间服务器一直正常，尝试了重装debian好几个版本，故障依旧。

**Reply by Jinnrry** [reply]
docker logs [容器ID] 看下日志呢

**Comment by Erick-shi** [comment]
你好，在 自己的虚拟机上部署不了吗？

**Comment by zzzyyyy123** [comment]
有很多垃圾邮件怎么禁止啊？

**Reply by Jinnrry** [reply]
目前支持2种防垃圾方案
1、配置文件中  spamFilterLevel设置为1或2，这个配置表示垃圾邮件过滤级别，0不过滤、1 spf dkim 校验均失败时过滤，2 spf校验不通过时过滤，这个设置是基于SMTP协议进行过滤的，如果对方服务器完全符合spf和dkim校验，那么是过滤不了的


2、基于机器学习的过滤方案，https://github.com/Jinnrry/PMail/blob/master/README_CN.md
